File: word_chain.scala
import scala.collection.mutable
import scala.io.StdIn._

class WordChain(n: Int, d: mutable.Set[String]) {
  private val alive = Array.fill(n)(true)
  private var bef = ""
  private var p = 0

  def wordChain(s: String): Unit = {
    if (d.contains(s) && (bef.isEmpty || bef.last == s.head) && s.last != 'z') {
      d -= s
      bef = s
    } else {
      alive(p) = false
      bef = ""
    }
    do {
      p = (p + 1) % n
    } while (!alive(p))
  }

  def aliveCount: Int = alive.count(identity)

  def isAlive(i: Int): Boolean = alive(i)
}

object Main extends App {
  val Array(n, k, m) = readLine().split(" ").map(_.toInt)
  val d = mutable.Set[String]()
  for (_ <- 0 until k) {
    d += readLine().trim
  }
  
  val wordChain = new WordChain(n, d)
  for (_ <- 0 until m) {
    wordChain.wordChain(readLine().trim)
  }

  println(wordChain.aliveCount)
  for (i <- 0 until n if wordChain.isAlive(i)) {
    println(i + 1)
  }
}

File: word_chain.js
class WordChain {
    N = 0;
    D = null;
    alive = null;
    bef = "";
    P = 0;

    constructor(n, d) {
        this.N = n;
        this.D = d;
        this.alive = new Array(n).fill(true);
    }

    wordChain(s) {
        if (this.D.has(s) && (this.bef === "" || this.bef[~-this.bef.length] === s[0]) && s[~-s.length] !== 'z') {
            this.D.delete(s);
            this.bef = s;
        } else {
            this.alive[this.P] = false;
            this.bef = "";
        }
        do {
            this.P = -~this.P % -~this.N;
        } while (!this.alive[this.P]);
    }
}


const lines = require("fs").readFileSync("/dev/stdin", "utf8").trim().split("\n");
const [N, K, M] = lines[0].split(" ").map(Number);
const D = new Set();
for (var k = 1; k <= K; k++) D.add(lines[k]);
const wordChain = new WordChain(N, D);
for (var m = 1; m <= M; m++) wordChain.wordChain(lines[K + m]);
console.log(wordChain.alive.filter(b => b).length)
for (var i = 0; i < N; i++) if (wordChain.alive[i]) console.log(-~i);